### PR TITLE
Create add to custom fields documentation

### LINKED ISSUE #10571: Add documentation outlining how to add a field to the checkout block, and how to process that in the checkout POST request.
There is no specific documentation to show how a developer can extend the checkout endpoint to receive additional data from a new inner block.

While there is https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-data.md#things-to-consider, this seems focused on _outputting_ data.

We should create a new document, or edit to the existing one linked above to include information on how this can be done.

### PR DESCRIPTION
**Pull request by tarhi-saad**
## What
This PR adds documentation outlining how to add a field to the checkout block and how to process that in the checkout POST request.

Fixes #10571

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

N/A

* [x] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

N/A

## WooCommerce Visibility

Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:

* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:

* [ ] This PR has a UI change and has been cross-browser tested at different viewport sizes on both the frontend and in the editor.
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

N/A

### CODE REVIEW

**Review by senadir — COMMENTED** [review]
I think, so that there isn't confusing between this doc and the work being done here https://github.com/woocommerce/woocommerce-blocks/discussions/11173 This document title or introduction should highlight well that you're building a field block, which is not the same as adding just a custom field.

**Comment by tarhi-saad** [comment]
I used `new field block` and `custom inner block` instead of `custom field` to avoid confusing between this doc and the work being done here https://github.com/woocommerce/woocommerce-blocks/discussions/11173

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+		"default": {
+			"remove": true,
+			"move": true
+		}
+	}
+}
+```
+
+- The [lock attribute](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-templates/#individual-block-locking) is an object that controls whether the block can be removed or moved. By default, the lock attribute is set to allow the block to be removed and moved. However, by modifying the lock attribute, you can “force” the block to be non-removable. For example, you can set both remove and move properties to false in order to prevent the block from being removed or moved.
+- The [parent attribute](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#parent) specifies the parent block that this block should be nested within it. It determines where the block will render. In our example, the block is a child of the `woocommerce/checkout-shipping-methods-block`. This means that the your block will be rendered within the `woocommerce/checkout-shipping-methods-block`.
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: Let's mention here that because it's a child of the shipping methods block, the block will not render if shipping is not required for the cart.

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
+# Extending the WooCommerce Checkout Block to Add a new Field Block
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: let's fix the casing here and make the title a little better.
```suggestion
# Add a new inner block containing a custom field to the WooCommerce Checkout Block
```

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+This documentation addresses the requirement for extending the WooCommerce checkout block to accept additional data from a new inner block.
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: This line doesn't seem all that useful, maybe we could reword it to:
```suggestion
This document describes how a developer can insert an input field into the Checkout block and have its value passed to the Store API so it's available when processing the checkout.
```

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+## Overview
+
+Developers can extend the checkout block to add new field blocks and process additional data through the checkout POST request. This involves leveraging the extensibility interfaces provided by WooCommerce Blocks, as demonstrated in [our tutorial](https://developer.woocommerce.com/2023/08/07/extending-the-woocommerce-checkout-block-to-add-custom-shipping-options/) for adding a "Not at home" shipping option.
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]:
```suggestion
Developers can extend the Checkout block to add new inner blocks and process additional data through the checkout POST request. This involves leveraging the extensibility interfaces provided by Gutenberg and WooCommerce Blocks. This is demonstrated in more detail in our tutorial: [Tutorial: Extending the WooCommerce Checkout Block
](https://developer.woocommerce.com/2023/08/07/extending-the-woocommerce-checkout-block-to-add-custom-shipping-options/).
```

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	}
+}
+```
+
+- The [lock attribute](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-templates/#individual-block-locking) is an object that controls whether the block can be removed or moved. By default, the lock attribute is set to allow the block to be removed and moved. However, by modifying the lock attribute, you can “force” the block to be non-removable. For example, you can set both remove and move properties to false in order to prevent the block from being removed or moved.
+- The [parent attribute](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#parent) specifies the parent block that this block should be nested within it. It determines where the block will render. In our example, the block is a child of the `woocommerce/checkout-shipping-methods-block`. This means that the your block will be rendered within the `woocommerce/checkout-shipping-methods-block`.
+
+### 3. Setting custom checkout data
+
+We can set the added field data to send it to the `wc/store/checkout` endpoint when processing orders using the function `setExtensionData`:
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md:51]: Let's go into more detail here about what `setExtensionData` is doing, and how we get it.

It's passed to inner blocks via props.

It updates the `extensionData` key of the `wc/store/checkout` _data store_.

This key is passed as part of the request body when POSTing to the checkout endpoint.

Should we show an example of getting this from the block's props? (The example is in the tutorial)

Comment (opr) [reply docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md:51]: Maybe we can also add screenshots of Redux Dev tool showing the data store before and after the `setExtensionData` call?

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+#### Parameters
+
+- namespace `string` - The namespace of your block.
+- key `string` - The key of your data.
+- value `any` - The value of your data.
+
+### 4. Processing the Checkout POST Request
+
+To process the added field data, extend the Store API. We can do this within the `initialize` of your `IntegrationInterface` class.
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: I think we should re-word this slightly:

```suggestion
To process the added field data, we'll need extend the Store API to tell it to expect additional data. A good place to start this process off is within the `initialize` function of our `IntegrationInterface` class.
```

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+					'namespace'       => self::IDENTIFIER,
+					'schema_callback' => [ 'Custom_Inner_Block_Extend_Store_Endpoint', 'extend_checkout_schema' ],
+					'schema_type'     => ARRAY_A,
+				]
+			);
+		}
+	}
+
+	/**
+	 * Register shipping workshop schema into the Checkout endpoint.
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: Consider rewording this description, it refers to the shipping workshop

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	 *
+	 * @return array Registered schema.
+	 *
+	 */
+	public static function extend_checkout_schema() {
+		return [
+            'Value_1'   => [
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md:204]: Let's fill in the values here so the code example works

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+                'readonly'    => // ... whether the field is readonly or not, this should be a boolean
+                'optional'    => // ... whether the field is optional or not, this should be a boolean
+            ],
+			// ... other values
+        ];
+	}
+}
+```
+
+## Conclusion
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md:283]: Before we conclude, we should have an example of how to _get_ this custom field's data while processing the checkout, for example in the `woocommerce_store_api_checkout_update_order_from_request` hook.

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+	/**
+	 * Extends the cart schema to include the custom-inner-block value.
+	 */
+	private function extend_store_api() {
+		Custom_Inner_Block_Extend_Store_Endpoint::init();
+	}
+}
+```
+
+- The `custom-inner-block-extend-store-endpoint.php` file: extends the [Store API](https://github.com/woocommerce/woocommerce-blocks/tree/trunk/src/StoreApi) and adds hooks to save and display your new field block instructions.
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: This doesn't save the data from the custom block anywhere by default, the developer would have to be explicit about this.

**Review by opr — CHANGES_REQUESTED** [review]
Thanks @tarhi-saad this is a good start, I've left some comments that we should address.

I'd also recommend following this documentation in your own example plugin and make sure all examples compile and achieve the desired result.

Thanks!

File: docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+Screenshots of Redux Dev tool showing the data store before and after the setExtensionData call:
+
+| Before | After |
+| ------ | ----- |
+|    <img width="713" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/14235870/948581f5-fdc2-4df1-963f-9aeb4b18b042">    |    <img width="723" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/14235870/ddc7dbe7-3fad-44cd-bd19-ce78bc49b951">   |
+
+### 4. Processing the Checkout POST Request
+
+To process the added field data, we'll need extend the Store API to tell it to expect additional data. A good place to start this process off is within the `initialize` function of our `IntegrationInterface` class.
```
Comment (opr) [inline docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: I checked on this recently, and actually the `IntegrationInterface` shouldn't be loaded on requests where blocks might not render (e.g. store API requests).

I'd therefore recommend creating a new class specifically for extending the Store API and not doing this as part of `IntegrationInterface`.

Comment (tarhi-saad) [reply docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: @opr, do you mean that we shouldn't use the following code since it implements `IntegrationInterface`?

- The `custom-inner-block-blocks-integration.php` file:

```php
use Automattic\WooCommerce\Blocks\Integrations\IntegrationInterface;

/**
 * Class for integrating with WooCommerce Blocks
 */
class Custom_Inner_Block_Blocks_Integration implements IntegrationInterface {
	// ... Some code here

	/**
	 * When called invokes any initialization/setup for the integration.
	 */
	public function initialize() {
		require_once __DIR__ . '/custom-inner-block-extend-store-endpoint.php';
		// ... Some code here: (e.g. init functions that registers scripts and styles, and other instructions)
		$this->extend_store_api();
	}

	/**
	 * Extends the cart schema to include the custom-inner-block value.
	 */
	private function extend_store_api() {
		Custom_Inner_Block_Extend_Store_Endpoint::init();
	}
}
```

We execute `$this->extend_store_api();` inside the `initialize` function. Is this the part that we should move to a new Class?

Comment (opr) [reply docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: > We execute `$this->extend_store_api();` inside the `initialize` function. Is this the part that we should move to a new Class?

exactly!

Comment (tarhi-saad) [reply docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md]: @opr, I didn't create a new Class for extending the Store API. I think we can use the existing `Custom_Inner_Block_Extend_Store_Endpoint` and initialize it on the plugin php file. I updated the description of our PHP files and added links to our documentation for more details.

**Review by opr — CHANGES_REQUESTED** [review]
Much better, thanks @tarhi-saad - just one last thing is to show where the user can invoke `Custom_Inner_Block_Extend_Store_Endpoint::init()` - maybe show an example of that happening on the `woocommerce_blocks_loaded` action?

Thanks!

**Comment by tarhi-saad** [comment]
Could you give this PR another look, @opr? 🙌

**Review by opr — APPROVED** [review]
Thanks @tarhi-saad this is looking good now! I appreciate you addressing the changes 🙏🏼

**Comment by tarhi-saad** [comment]
> Thanks @tarhi-saad this is looking good now! I appreciate you addressing the changes 🙏🏼

Thank you, @opr, for your feedback! I learned a lot from it! 🙌
